source/common/ssl/ssl_socket.cc

```diff
@@ -384,31 +416,83 @@ std::string SslSocket::subjectLocalCertificate() const {
   return getSubjectFromCertificate(cert);
 }
 
+namespace {
+SslSocketFactoryStats generateStats(const std::string& prefix, Stats::Scope& store) {
+  return {
+      ALL_SSL_SOCKET_FACTORY_STATS(POOL_COUNTER_PREFIX(store, prefix + "_ssl_socket_factory."))};
+}
+} // namespace
+
 ClientSslSocketFactory::ClientSslSocketFactory(ClientContextConfigPtr config,
                                                Ssl::ContextManager& manager,
                                                Stats::Scope& stats_scope)
-    : manager_(manager), stats_scope_(stats_scope), config_(std::move(config)),
-      ssl_ctx_(manager_.createSslClientContext(stats_scope_, *config_)) {}
+    : manager_(manager), stats_scope_(stats_scope), stats_(generateStats("client", stats_scope)),
+      config_(std::move(config)),
+      ssl_ctx_(manager_.createSslClientContext(stats_scope_, *config_)) {
+  if (config_->getSecretProvider()) {
```

Comment:
can we add addUpdateCallback to ContextConfig itself, rather than checking here. ContextConfig might have other secrets (session ticket, validation context) in future.